lenskit-slim/src/main/java/org/lenskit/slim/LinearRegressionHelper.java

```diff
@@ -0,0 +1,143 @@
+        Long2ObjectMap<Long2DoubleMap> mapT = new Long2ObjectOpenHashMap<>();
+
+        for (Map.Entry<Long, Long2DoubleMap> rowEntry : matrix.entrySet()) {
+            long rowNum = rowEntry.getKey();
+            Long2DoubleMap rowValue = rowEntry.getValue();
+            for (Map.Entry<Long, Double> entry : rowValue.entrySet()) {
+                long colNum = entry.getKey();
+                double Value = entry.getValue();
+                Long2DoubleMap rowOfMapT = mapT.get(colNum);
+                if (rowOfMapT == null) rowOfMapT = new Long2DoubleOpenHashMap();
+                rowOfMapT.put(rowNum, Value);
+                mapT.put(colNum, rowOfMapT);
+            }
+        }
+        return mapT;
+    }
+
+    /**
+     * add two vectors
+     * @param a
+     * @param b
+     * @return
+     */
+    public static Long2DoubleMap addVectors(Long2DoubleMap a, Long2DoubleMap b) {
+        //Long2DoubleMap sumOfTwoVectors = Vectors.combine(a, b, 1.0, 0.0);
+        Long2DoubleMap sumOfTwoVectors = new Long2DoubleOpenHashMap(a);
+        for (Map.Entry<Long, Double> e : b.entrySet()) {
+            long key = e.getKey();
+            double value = e.getValue();
+            double sum = sumOfTwoVectors.get(key) + value;
+            sumOfTwoVectors.put(key, sum);
+        }
+        return sumOfTwoVectors;
+
+    }
+
+    /**
+     * element-wise multiplication of two vectors.
+     * @param a
+     * @param b
+     * @return
+     */
+    public static Long2DoubleMap multiply(Long2DoubleMap a, Long2DoubleMap b) {
+        Long2DoubleMap productOfTwoVectors = new Long2DoubleOpenHashMap();
+        LongOpenHashBigSet aSet = new LongOpenHashBigSet(a.keySet());
+        LongOpenHashBigSet bSet = new LongOpenHashBigSet(b.keySet());
+        aSet.retainAll(bSet); // intersection of two key sets
+        //productOfTwoVectors.defaultReturnValue(0.0);
+        //b.forEach((k, v) -> productOfTwoVectors.merge(k, v, (oldVal, newVal) -> oldVal * newVal));
+        for (long key : aSet) {
+            //double prod = a.getOrDefault(key, 0.0)*b.getOrDefault(key, 0.0);
+            double prod = a.get(key)*b.get(key);
+            productOfTwoVectors.put(key, prod);
+        }
+        return productOfTwoVectors;
+    }
+
+//    /**
+//     * multiply vector by a real value
+//     * @param a
+//     * @param c
+//     * @return
+//     */
+//    public static Long2DoubleMap multiply(Long2DoubleMap a, double c) {
+//        Long2DoubleMap scalingVector = new Long2DoubleOpenHashMap();
+//        a.forEach((k, v) -> scalingVector.put((long) k, (v*c)));
+//        return scalingVector;
+//    }
+
+    /**
+     * return a new map with all values in input map but not equal to the given value
+     * @param a input map
+     * @param value a value needs to be filtered out
+     * @return
+     */
+    public static Long2DoubleMap filterValues(Long2DoubleMap a, double value) {
```

Comment:
We might want to introduce a customizable epsilon here? `Scalars.isZero` is pretty restrictive by default.